pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/StringPredicateFilterOptimizer.java

```diff
@@ -0,0 +1,127 @@
+
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.request.Expression;
+import org.apache.pinot.common.request.ExpressionType;
+import org.apache.pinot.common.request.Function;
+import org.apache.pinot.common.request.PinotQuery;
+import org.apache.pinot.pql.parsers.pql2.ast.FilterKind;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+
+
+/**
+ * Given two column names 'strColumn1' and 'strColumn1', CalciteSqlParser.queryRewrite will turn WHERE and HAVING
+ * expressions of form "strColumn1 <operator> strColumn2" into "MINUS(strColumn1,strColumn2) <operator> 0" regardless
+ * of the column datatype. The resulting query will fail to evaluate since the MINUS operator does not work with the
+ * STRING column type. This class rewrites expressions of form "MINUS(strColumn1,strColumn2) <operator> 0" to
+ * "STRCMP(strColumn1, strColumn2) <operator> 0" to fix the issue.
+ *
+ * Currently, rewrite phase (see CalciteSqlParser.queryRewrite) does not have access to schema; hence, we need to again
+ * rewrite MINUS(strColumn1, strColumn2) into STRCMP(strColumn1, strColumn2). At some point, we should merge query
+ * rewrite phase with optimizer phase to avoid such issues altogether.
+ */
+public class StringPredicateFilterOptimizer implements StatementOptimizer {
+  private static final String MINUS_OPERATOR_NAME = "MINUS";
+  private static final String COMPARE_OPERATOR_NAME = "STRCMP";
+
+  @Override
+  public void optimize(PinotQuery query, @Nullable TableConfig tableConfig, @Nullable Schema schema) {
+    Expression filter = query.getFilterExpression();
+    if (filter != null && schema != null) {
+      optimizeExpression(filter, schema);
+    }
+
+    Expression expression = query.getHavingExpression();
+    if (expression != null && schema != null) {
+      optimizeExpression(expression, schema);
+    }
+  }
+
+  /** Traverse an expression tree to replace MINUS function with COMPARE if function operands are STRING. */
+  private static void optimizeExpression(Expression expression, Schema schema) {
+    ExpressionType type = expression.getType();
+    if (type != ExpressionType.FUNCTION) {
+      // We have nothing to rewrite if expression is not a function.
+      return;
+    }
+
+    Function function = expression.getFunctionCall();
+    String operator = function.getOperator();
+    List<Expression> operands = function.getOperands();
+    FilterKind kind = FilterKind.valueOf(operator);
+    switch (kind) {
+      case AND:
+      case OR: {
+        optimizeExpression(operands.get(0), schema);
+        break;
+      }
+      case EQUALS:
+      case NOT_EQUALS:
+      case GREATER_THAN:
+      case GREATER_THAN_OR_EQUAL:
+      case LESS_THAN:
+      case LESS_THAN_OR_EQUAL:
+      case IS_NULL:
+      case IS_NOT_NULL: {
+        replaceMinusWithCompareForStrings(operands.get(0), schema);
```

Comment:
We should put this in the default to cover all cases. E.g. when user explicitly put (str1 - str2) instead of converted from the parser

Reply:
To simplify, we can use if
```suggestion
    if (kind == AND || kind == OR) {
      for ...
    } else {
      replaceMinusWithCompareForStrings(operands.get(0), schema);
    }
```

Reply:
Moved to "default. Let's leave it as "switch" though as switch tends to perform better with a set of options as compared to if and probably easier to modify in future as well.